test/animationinformation_test.cpp

```diff
@@ -0,0 +1,371 @@
+		auto renderingData = dynamic_cast<RenderingData *>(x);
+		if (renderingData != nullptr) {
+			gfProgressToNextGameTick = renderingData->_fProgressToNextGameTick;
+			EXPECT_EQ(GetFrameToUseForPlayerRendering(pPlayer), renderingData->_ExpectedRenderingFrame)
+			    << std::fixed << std::setprecision(2)
+			    << "ProgressToNextGameTick: " << renderingData->_fProgressToNextGameTick
+			    << " AnimFrame: " << pPlayer->_pAnimFrame
+			    << " AnimCnt: " << pPlayer->_pAnimCnt
+			    << " GameTick: " << currentGameTick;
+		}
+	}
+	for (TestData *x : vecTestData) {
+		delete x;
+	}
+}
+
+TEST(AnimationInformation, AttackSwordWarrior) // ProcessAnimationPending should be considered by distribution logic
+{
+	RunAnimationTest(16, 0, AnimationFlags::ProcessAnimationPending, 0, 9,
+	    {
+	        new GameTickData(2, 0), // ProcessPlayer directly after StartAttack (in same GameTick). So we don't see any rendering before.
+	        new RenderingData(0.0f, 1),
+	        new RenderingData(0.3f, 1),
+	        new RenderingData(0.6f, 1),
+	        new RenderingData(0.8f, 1),
+	        new GameTickData(3, 0),
+	        new RenderingData(0.0f, 2),
+	        new RenderingData(0.3f, 2),
+	        new RenderingData(0.6f, 2),
+	        new RenderingData(0.8f, 3),
+	        new GameTickData(4, 0),
+	        new RenderingData(0.0f, 3),
+	        new RenderingData(0.3f, 3),
+	        new RenderingData(0.6f, 3),
+	        new RenderingData(0.8f, 4),
+	        new GameTickData(5, 0),
+	        new RenderingData(0.0f, 4),
+	        new RenderingData(0.3f, 4),
+	        new RenderingData(0.6f, 5),
+	        new RenderingData(0.8f, 5),
+	        new GameTickData(6, 0),
+	        new RenderingData(0.0f, 5),
+	        new RenderingData(0.3f, 5),
+	        new RenderingData(0.6f, 6),
+	        new RenderingData(0.8f, 6),
+	        new GameTickData(7, 0),
+	        new RenderingData(0.0f, 6),
+	        new RenderingData(0.3f, 7),
+	        new RenderingData(0.6f, 7),
+	        new RenderingData(0.8f, 7),
+	        new GameTickData(8, 0),
+	        new RenderingData(0.0f, 7),
+	        new RenderingData(0.3f, 8),
+	        new RenderingData(0.6f, 8),
+	        new RenderingData(0.8f, 8),
+
+	        new GameTickData(9, 0), // After this GameTick, the Animation Distribution Logic is disabled
+	        new RenderingData(0.1f, 9),
+	        new GameTickData(10, 0),
+	        new RenderingData(0.4f, 10),
+	        new GameTickData(11, 0),
+	        new RenderingData(0.4f, 11),
+	        new GameTickData(12, 0),
+	        new RenderingData(0.3f, 12),
+	        new GameTickData(13, 0),
+	        new RenderingData(0.0f, 13),
+	        new GameTickData(14, 0),
+	        new RenderingData(0.6f, 14),
+	        new GameTickData(15, 0),
+	        new RenderingData(0.6f, 15),
+	        new GameTickData(16, 0),
+	        new RenderingData(0.6f, 16),
+	        // Animation stopped cause PM_DoAttack would stop the Animation "if (plr[pnum]._pAnimFrame == plr[pnum]._pAFrames) {"
+	    });
```

Comment:
I added a comment. Does this help?
Have you other suggestions what can be changed to improve the unittest?

Reply:
Comments to explain code should be avoided as much as possible. The problem I had here is that the code itself is really hard to reason about, and thus the code itself should be improved instead.

> Have you other suggestions what can be changed to improve the unittest?

Not really no... as I said this change has got a bit out of hand for me, and it is tough to suggest anything if I don't clearly understand everything going on.

I'm just giving feedback as someone "looking from a distance", basically, so take it with a grain of salt.

Reply:
I would also say it's best if the code can be understand without comments. that's also why good naming is really important.
But with increasing complexity it's not always possible to avoid comments.
I wanted to leave something in the Sourcecode that other people could help in this regard.

That's also why we need to read documentations of our used language or framework.
If you don't mind I would close this discussion as I would need input how I can help with this matter.
The bigger picture and decision decision we have to take is described in my last [comment](https://github.com/diasurgical/devilutionX/pull/1539#issuecomment-821762887).

But I'm open to any future feedback, as I want a good solution for this project.

Reply:
> If you don't mind I would close this discussion as I would need input how I can help with this matter.

I don't mind at all. It was added to see if you could come up with an alternative yourself.

Maybe other folks can provide more guidance than me on this particular one.

As i said, i don't feel qualified to properly review a good portion of this code.